kratos/containers/distributed_csr_matrix.h

```diff
@@ -0,0 +1,773 @@
+                    k++;
+                }
+            }
+            mDiagBlock.index1_data()[counter+1] = row_begin + k;
+            counter++;
+        }
+        mDiagBlock.SetColSize(GetColNumbering().LocalSize());
+        
+#ifdef KRATOS_DEBUG
+        mDiagBlock.CheckColSize();
+#endif
+
+        //ensure columns are ordered
+        for(IndexType i = 0; i<nlocal_rows; ++i){
+            IndexType row_begin = mDiagBlock.index1_data()[i];
+            IndexType row_end = mDiagBlock.index1_data()[i+1];
+            std::sort(mDiagBlock.index2_data().begin() + row_begin ,mDiagBlock.index2_data().begin()+row_end); 
+        }
+
+
+        //*******************************
+        //construct offdiagonal block
+        
+        //store off-diagonal block
+        mOffDiagBlock.reserve(nlocal_rows, offdiag_nnz); 
+        counter = 0;
+        for(const auto& entries : local_graph){
+            unsigned int k = 0;
+            IndexType row_begin = mOffDiagBlock.index1_data()[counter];
+            for(auto global_j : entries){
+                if( ! GetColNumbering().IsLocal(global_j)){
+                    IndexType local_j = GetOffDiagLocalId(global_j);
+                    mOffDiagBlock.index2_data()[row_begin+k] = local_j;
+                    mOffDiagBlock.value_data()[row_begin+k] = 0.0;
+                    k++;
+                }
+            }
+            mOffDiagBlock.index1_data()[counter+1] = row_begin + k;
+            counter++;
+        }
+        mOffDiagBlock.SetColSize(mOffDiagonalLocalIds.size());
+
+#ifdef KRATOS_DEBUG
+        mOffDiagBlock.CheckColSize();
+#endif
+
+        //ensure columns are ordered
+        for(IndexType i = 0; i<nlocal_rows; ++i){
+            IndexType row_begin = mOffDiagBlock.index1_data()[i];
+            IndexType row_end = mOffDiagBlock.index1_data()[i+1];
+            std::sort(mOffDiagBlock.index2_data().begin() + row_begin ,mOffDiagBlock.index2_data().begin()+row_end); 
+        }
+
+        PrepareNonLocalCommunications(rSparseGraph);
+
+        //mount importer for SpMV calculations
+
+        auto pimporter = Kratos::make_unique<DistributedVectorImporter<double,IndexType>>(GetComm(),mOffDiagonalGlobalIds, GetColNumbering()); 
+        mpVectorImporter.swap(pimporter);
+    }
+
+    /// Destructor.
+    virtual ~DistributedCsrMatrix(){}
+
+    /// Assignment operator. TODO: decide if we do want to allow it
+    DistributedCsrMatrix& operator=(DistributedCsrMatrix const& rOther)=delete;
+    // {
+    //     this->AddEntries(rOther.GetGraph());
+    //     return *this;
+    // }
+
+    ///@}
+    ///@name Operators
+    ///@{
+    void Clear()
+    {
+
+    }
+
+    inline const DistributedNumbering<IndexType>& GetRowNumbering() const
+    {
+        return *mpRowNumbering;
+    }
+    
+    inline const DistributedNumbering<IndexType>& GetColNumbering() const
+    {
+        return *mpColNumbering;
+    }
+
+    void SetValue(const TDataType value)
```

Comment:
Same comment as in the `csr_matrix`.

Reply:
Fine. Same suggestion about creating a new `SetToZero` in both matrices.

Reply:
OK, I don't have a strong reason against it.